---
title: 'Class 16: Genome annotation and the interpretation of gene lists'
author: "Caitlin Johnson"
date: "2/27/2020"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Section 1: Differential Expression Analysis
```{r}
library(DESeq2)
```

Let's load our data files
The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703
The authors report on differential analysis of lung fibroblasts in response to loss of the developmental transcription factor HOXA1. Their results and others indicate that HOXA1 is required for lung fibroblast and HeLa cell cycle progression. In particular their analysis show that "loss of HOXA1 results in significant expression level changes in thousands of individual transcripts, along with isoform switching events in key regulators of the cell cycle". For our session we have used their Sailfish gene-level estimated counts and hence are restricted to protein-coding genes only.
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"
#Import metadata and countdata
colData = read.csv(metaFile, row.names = 1)
head(colData)
countData = read.csv(countFile, row.names = 1)
head(countData)
```
There are `r nrow(countData)` genes in this dataset and `r ncol(countData)` experiments (samples)
```{r}

```

We need the `countData` and `colData` files to match up so we need to remove the `countData$length`
Q1: Complete the code below to remove the troulesome first column from `countData`
